<launch>
    <!-- Set use_sim_time true to generate trajectory bag file in same time domain -->
    <param name="/use_sim_time" value="true"/>

    <!-- Path of a rosbag file -->
    <arg name="bag_path" default="$(find path_recorder)/../bag/kitti_2011_09_30_drive_0027_synced"/>
   <!-- For Velodyne VLP-16 -->
   <param name="scan_line" value="16" />
    

   <!-- For Velodyne HDL-32 
   <param name="scan_line" value="32" />
   -->

   <!-- For Velodyne HDL-64 -->
   <param name="scan_line" value="64" />
   

    <!-- Run Floam -->
    <group>
          <!--- Sim Time -->
    <param name="scan_period" value="0.1" />

    <param name="vertical_angle" type="double" value="2.0" />
    <param name="map_resolution" type="double" value="0.2" />
    <param name="max_dis" type="double" value="90.0" />
    <param name="min_dis" type="double" value="3.0" />
    <!--- -->
    <node pkg="floam" type="floam_odom_estimation_node" name="floam_odom_estimation_node" output="screen"/>
    <node pkg="floam" type="floam_laser_processing_node" name="floam_laser_processing_node" output="screen"/>
    <!-- <node pkg="floam" type="floam_laser_mapping_node" name="floam_laser_mapping_node" output="screen"/> -->

    <node pkg="tf" type="static_transform_publisher" name="word2map_tf"  args="0 0 0 0 0 0 /world /map 10" />
    
    <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="trajectory_server_loam" ns="base_link" >
        <param name="/target_frame_name" value="world" />
        <param name="/source_frame_name" value="base_link" />
        <param name="/trajectory_update_rate" value="10.0" />
        <param name="/trajectory_publish_rate" value="10.0" />
    </node>

    <arg name="rviz" default="true" />
    <group if="$(arg rviz)">
        <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find floam)/rviz/floam_velodyne.rviz" />
    </group>
    </group>


	<!-- Remap point cloud topic -->
	<remap from="points_raw" to="velodyne_points" />
	<!-- Play kitti rosbag file -->
    <node pkg="rosbag" type="play" name="kitti" args="--clock $(arg bag_path)/kitti.bag" output="log" />

	<!-- Launch path_recorder to save a trajectory from KISS-ICP -->
    <node pkg="path_recorder" type="path_recorder" name="kiss_recorder" output="screen">
        <param name="algorithm_name" type="string" value="f_loam" />
        <param name="path_topic" type="string" value="/base_link/trajectory" />
        <param name="save_to" type="string" value="$(arg bag_path)" />
    </node>
</launch>
